### PR TITLE
Add interface for `AptosClient` to get `Object<T>`

### PR DESCRIPTION
**Pull request by SamuelQZQ**
### Description

Add interface for `AptosClient` to get `Object<T>`.

### Test Plan
I tested locally. Do not know use what address to do the E2E test.

### CODE REVIEW

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 1]
```diff
@@ -48,6 +48,14 @@ test("gets genesis resources", async () => {
   expect(accountResource).toBeDefined();
 });
 
+// test("gets object", async () => {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:51]: yea.. you would need to create an object and then query it to test it. can do something like that
```
const alice = new AptosAccount();
const faucetClient = getFaucetClient();
await faucetClient.fundAccount(alice.address(), 100000000);

const provider = new Provider({
        fullnodeUrl: NODE_URL,
        indexerUrl: NODE_URL,
      });
const aptosToken = new AptosToken(provider)

const txn = await client.waitForTransactionWithResult(
        await aptosToken.createCollection(alice, "Alice's simple collection","AliceCollection", "https://aptos.dev", 5),
        { checkSuccess: true },
      );
const objectAddress = txn.changes[0].address // should be the new object address
```

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 2]
```diff
@@ -48,6 +48,14 @@ test("gets genesis resources", async () => {
...
+//   const client = new AptosClient(NODE_URL);
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:52]: can we also test for invalid address error? i.e expect().toThrow()

**Review by 0xmaayan — COMMENTED** [review]
Thanks! great addition! Left some comments for a way to test the new query

File: ecosystem/typescript/sdk/src/providers/aptos_client.ts [hunk 3]
```diff
@@ -233,6 +233,25 @@ export class AptosClient {
...
+   * @param query.ledgerVersion Specifies ledger version of transactions. By default latest version will be used
+   * @returns 0x1::object::Object resource
+   */
+    @parseApiError
+    async getObjectResource(
+      address: MaybeHexString,
+      query?: { ledgerVersion?: AnyNumber },
+    ): Promise<Gen.MoveResource> {
+      const resources = await this.getAccountResources(address, query);
+      if(resources.length !== 2 || resources.find((r) => r.type === "0x1::object::ObjectCore") === undefined) {
```
Comment (gregnazario) [inline ecosystem/typescript/sdk/src/providers/aptos_client.ts:248]: Let's clean this up.

The resources can be > 2 and also I'm not sure there's a guarantee that 0x1::object::ObjectCore will be in the first page of getAccountResources.

Let's instead use `getAccountResource` and get specifically the `0x1::object::ObjectCore` resource.

Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts:248]: @gregnazario Sorry, I think what we want from this `getObjectResource` interface is not the `0x1::object::ObjectCore` resource. What we need is the `0x1::object::Object<T0>` resource. With the `address` as input arguments, we do not know the `T0`. So we cannot use `getAccountResource` to get the `Object` resource. Any idea?

You can see this example:
https://fullnode.testnet.aptoslabs.com/v1/accounts/0xb584c623d565f87cc3c20df8fad99b9a1d2783d00e1ff00236faa05190a7ae4f/resources

Comment (gregnazario) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts:248]: I see, there's a misunderstanding of how Objects work.

An object would be all of the applicable resources, and could be extended.

You could have an object as a resource group with 3 different resources say:
1. Cat
2. Dog
3. Monkey

You could access that address (the same object) as one of three types:
1. Object<Cat>
2. Object<Dog>
3. Object<Monkey>

This allows the flexibility of bundling these together

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts (hunk 1 above)
Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts]: Added

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts (hunk 2 above)
Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts]: Now it's same with `getAccountResource` interface. The invalid account resource error is removed.

File: ecosystem/typescript/sdk/src/providers/aptos_client.ts (hunk 3 above)
Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts]: Changed.

File: ecosystem/typescript/sdk/CHANGELOG.md [hunk 4]
```diff
@@ -4,7 +4,10 @@ All notable changes to the Aptos Node SDK will be captured in this file. This ch
 
 ## Unreleased
 
+- Add `getObjectResource` interface for `AptosClient` to get `Object<T>`
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/CHANGELOG.md:7]:
```suggestion
- Add `getObjectResource` interface for `AptosClient` to get `Object<T>` resource
```

File: ecosystem/typescript/sdk/src/providers/aptos_client.ts [hunk 5]
```diff
@@ -233,6 +233,22 @@ export class AptosClient {
     return data;
   }
 
+  /**
+   * Queries a 0x1::object::Object resource associated with given address by resource type
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/providers/aptos_client.ts:237]: tbh, I am not completely sure what are the benefits with this function? it is exactly the same as `getAccountResource` but just that the first argument name is different (objectAddress vs accountAddress)

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 6]
```diff
@@ -48,6 +49,33 @@ test("gets genesis resources", async () => {
...
+  const provider = new Provider({
+    fullnodeUrl: NODE_URL,
+    indexerUrl: NODE_URL,
+  });
+  const aptosToken = new AptosToken(provider)
+
+  const txn = await client.waitForTransactionWithResult(
+    await aptosToken.createCollection(alice, "Alice's simple collection", "AliceCollection", "https://aptos.dev", 5, {
+      royaltyNumerator: 10,
+      royaltyDenominator: 10,
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:67]:
```suggestion

```

File: ecosystem/typescript/sdk/src/providers/aptos_client.ts (hunk 5 above)
Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts:237]: @0xmaayan Yes, user can just use the `getAccountResource`  to get the object. Maybe no need to add this function. Just document somewhere to let user know how to get the object with AptosClient.

Comment (0xmaayan) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts:237]: agreed. maybe adding a test that "gets an object" so users would have an example on how they can get an object

Comment (SamuelQZQ) [reply ecosystem/typescript/sdk/src/providers/aptos_client.ts]: @0xmaayan changed. How could I request the CI?

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 7]
```diff
@@ -48,6 +49,30 @@ test("gets genesis resources", async () => {
   expect(accountResource).toBeDefined();
 });
 
+test("gets object", async () => {
+  const alice = new AptosAccount();
+  const faucetClient = getFaucetClient();
+  await faucetClient.fundAccount(alice.address(), 100000000);
+
+  const client = new AptosClient(NODE_URL);
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:57]:
```suggestion
```

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 8]
```diff
@@ -48,6 +49,30 @@ test("gets genesis resources", async () => {
...
+  const provider = new Provider({
+    fullnodeUrl: NODE_URL,
+    indexerUrl: NODE_URL,
+  });
+  const aptosToken = new AptosToken(provider)
+
+  const txn = await client.waitForTransactionWithResult(
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:64]:
```suggestion
  const txn = await provider.waitForTransactionWithResult(
```

File: ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts [hunk 9]
```diff
@@ -48,6 +49,30 @@ test("gets genesis resources", async () => {
...
+    await aptosToken.createCollection(alice, "Alice's simple collection", "AliceCollection", "https://aptos.dev"),
+    { checkSuccess: true },
+  );
+  const objectAddress = ((txn as Gen.UserTransaction).changes[0] as WriteSetChange_WriteResource).address // should be the new object address
+  
+  const object = await client.getAccountResource(objectAddress, "0x4::aptos_token::AptosCollection");
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/tests/e2e/aptos_client.test.ts:70]:
```suggestion
  const object = await provider.getAccountResource(objectAddress, "0x4::aptos_token::AptosCollection");
```

**Comment by SamuelQZQ** [comment]
> @SamuelQZQ lint is failing, please run `pnpm fmt` and `pnpm lint`. also test is failing, a fix for that is here #9341

Changed. And merged main branch. Should the test be fixed after the merge?

**Review by 0xmaayan — APPROVED** [review]

**Review by gregnazario — APPROVED** [review]

**Pull request merged**